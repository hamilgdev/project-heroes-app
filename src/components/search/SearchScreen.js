import queryString from "query-string";
import { useMemo } from "react";
import { useLocation, useNavigate } from "react-router-dom";
import { getHeroByName } from "../../helpers/getHeroByName";
import { useForm } from "../../hooks/useForm";
import HeroCard from "../hero/HeroCard";

const SearchScreen = () => {
  const navigate = useNavigate();
  const location = useLocation();

  const { q = "" } = queryString.parse(location.search);

  const [formValues, handleInputChange] = useForm({ searchText: q });
  const { searchText } = formValues;

  const heroesFilted = useMemo(() => getHeroByName(q), [q]);

  const handleSearch = (e) => {
    e.preventDefault();
    navigate(`?q=${searchText}`);
  };

  return (
    <section className="my-4">
      <h1 className="text-xl text-gray-600 text-center">
        <strong>Search Hero</strong> | 🦸‍♀️
      </h1>
      <form
        className="flex gap-4 my-4 flex-col sm:flex-row max-w-md m-auto text-center"
        onSubmit={handleSearch}
      >
        <input
          className="p-2 rounded-lg w-full bg-slate-50"
          type="text"
          placeholder="Iron man, Thor..."
          onChange={handleInputChange}
          name="searchText"
          value={searchText}
        />
        <button
          className="py-2 px-4 rounded-lg text-sky-600 bg-sky-300 | hover:bg-sky-600 hover:text-sky-300"
          type="submit"
        >
          Search
        </button>
      </form>

      {!heroesFilted.length && (
        <div className="p-4 bg-sky-200 rounded-md w-full">
          <p className="text-sky-600">Without result</p>
        </div>
      )}

      <div className="flex items-center justify-center flex-col gap-4 | sm:grid sm:justify-items-center sm:grid-cols-3 md:grid-cols-2 sm:justify-items-stretch xl:grid-cols-3">
        {heroesFilted.map((hero) => (
          <HeroCard key={hero.id} {...hero} />
        ))}
      </div>
    </section>
  );
};

export default SearchScreen;
