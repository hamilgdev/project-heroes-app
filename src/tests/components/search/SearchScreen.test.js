import { mount } from "enzyme";
import { MemoryRouter } from "react-router-dom";
import SearchScreen from "../../../components/search/SearchScreen";

describe("Tests in <SearchScreen />", () => {
  test("should display correctly by default", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search"]}>
        <SearchScreen />
      </MemoryRouter>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find("h1 strong").text().trim()).toBe("Search Hero");
  });

  test("should display to Batman and the input with the value of the queryString", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search?q=batman"]}>
        <SearchScreen />
      </MemoryRouter>
    );

    expect(wrapper.find("input").prop("value")).toBe("batman");
  });

  test("should display without result if hero no exist", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search?q=boyacoman"]}>
        <SearchScreen />
      </MemoryRouter>
    );

    expect(wrapper.find("div p").text().trim()).toBe("Without result");
  });
});
