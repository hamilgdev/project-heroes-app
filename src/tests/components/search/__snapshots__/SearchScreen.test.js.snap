// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <SearchScreen /> should display correctly by default 1`] = `
<section
  className="my-4"
>
  <h1
    className="text-xl text-gray-600 text-center"
  >
    <strong>
      Search Hero
    </strong>
     | 🦸‍♀️
  </h1>
  <form
    className="flex gap-4 my-4 flex-col sm:flex-row max-w-md m-auto text-center"
    onSubmit={[Function]}
  >
    <input
      className="p-2 rounded-lg w-full bg-slate-50"
      name="searchText"
      onChange={[Function]}
      placeholder="Iron man, Thor..."
      type="text"
      value=""
    />
    <button
      className="py-2 px-4 rounded-lg text-sky-600 bg-sky-300 | hover:bg-sky-600 hover:text-sky-300"
      type="submit"
    >
      Search
    </button>
  </form>
  <div
    className="p-4 bg-sky-200 rounded-md w-full"
  >
    <p
      className="text-sky-600"
    >
      Without result
    </p>
  </div>
  <div
    className="flex items-center justify-center flex-col gap-4 | sm:grid sm:justify-items-center sm:grid-cols-3 md:grid-cols-2 sm:justify-items-stretch xl:grid-cols-3"
  />
</section>
`;
